chart/templates/deployment.yaml

```diff
@@ -105,7 +110,7 @@ spec:
       securityContext:
         runAsNonRoot: true
       serviceAccountName: {{ include "chart.fullname" . }}-controller-manager
-      terminationGracePeriodSeconds: 10
+      terminationGracePeriodSeconds : {{ .Values.controller.terminationGracePeriodSeconds }}
```

Comment:
Given the pre-delete hook and owner references handle all the Kubernetes object deletions, is it possible to get rid of the config map by using this grace period and a `preStop` hook for the in-memory-only leases?

The tricky part is the pod needs to know the difference between the deployment getting deleted (revoke everything) and any other reason a pod could be deleted (do nothing). One signal we could send that doesn't require any RBAC permissions (and thus avoids a race condition with the RBAC being deleted) is an annotation that's projected into the pod via the volume downward API. The pre-delete hook could update that annotation on all the operator pods to say they should revoke everything on shutdown. WDYT?

Reply:
> a preStop hook for the in-memory-only leases

https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/. Looks like the `preStop` is a command. I'm not sure how to use the `preStop` hook to revoke in-memory-only leases of the running process/command `/vault-secrets-operator` though they are all in the same container. I wonder if you meant using it to signal `/vault-secrets-operator` to revoke the leases.

> One signal we could send that doesn't require any RBAC permissions (and thus avoids a race condition with the RBAC being deleted) is an annotation that's projected into the pod via the volume downward API. The pre-delete hook could update that annotation on all the operator pods to say they should revoke everything on shutdown

Using the annotation is a good idea. I wonder if there is a concept of watchers for annotations. I will research a bit on that.

Reply:
For the `preStop` command, we could do something like `kill -s HUP 1`. Then the operator can listen for that HUP signal and know that it should check the annotation value and start revoking if the deployment is being deleted: https://pkg.go.dev/os/signal#example-Notify

For the annotation, the downward API can project a specific annotation into a file within the pod: https://kubernetes.io/docs/concepts/workloads/pods/downward-api/#downwardapi-fieldRef. From there, you just need to know when to check the file. [fsnotify](https://github.com/fsnotify/fsnotify) is a good library for that generally, but I don't think it would be required in this case, because we can just listen for that HUP signal from the `preStop` hook.

Reply:
I just want to stress as well I'm not totally sure this will work! And I'm not totally confident it would turn out better/simpler. Just exploring ideas :-)

Reply:
> For the `preStop` command, we could do something like `kill -s HUP 1`. Then the operator can listen for that HUP signal and know that it should check the annotation value and start revoking if the deployment is being deleted: https://pkg.go.dev/os/signal#example-Notify

Would that require enabling shared process namespace?

> 
> For the annotation, the downward API can project a specific annotation into a file within the pod: https://kubernetes.io/docs/concepts/workloads/pods/downward-api/#downwardapi-fieldRef. From there, you just need to know when to check the file. [fsnotify](https://github.com/fsnotify/fsnotify) is a good library for that generally, but I don't think it would be required in this case, because we can just listen for that HUP signal from the `preStop` hook.

I like the idea of using inotify here, but I would hesitate to add it to the Operator especially when it comes to cross platform compatibility.

Reply:
In this case, I think best effort should be good enough. The token revocation is already best effort.

Reply:
@tomhjp. I found "updating the pod annotations will replace the pod" info on a random source. Sorry I should have tested it prior. Thanks for the example.